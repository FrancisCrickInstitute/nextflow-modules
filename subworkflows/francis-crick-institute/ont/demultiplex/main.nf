include { WISECONDORX_CONVERT } from '../../../modules/nf-core/wisecondorx/convert/main'
include { WISECONDORX_PREDICT } from '../../../modules/nf-core/wisecondorx/predict/main'

workflow ONT_DEMULTIPLEX {

    take:
    val_dorado_auto_model // Specify a short code like HAC or SUP
    val_dorado_model      // Specify an exact model which will override the previous

    // ch_bam          // channel: [ val(meta), path(bam), path(bai) ]
    // ch_fasta        // channel: [ val(meta2), path(fasta) ]
    // ch_fai          // channel: [ val(meta3), path(fai) ]
    // ch_ref          // channel: [ val(meta4), path(reference) ]
    // ch_blacklist    // channel: [ val(meta5), path(blacklist) ]

    main:

    ch_versions = Channel.empty()


    emit:
    // aberrations_bed = WISECONDORX_PREDICT.out.aberrations_bed   // channel: [ val(meta), path(bed) ]
    // bins_bed        = WISECONDORX_PREDICT.out.bins_bed          // channel: [ val(meta), path(bed) ]
    // segments_bed    = WISECONDORX_PREDICT.out.segments_bed      // channel: [ val(meta), path(bed) ]
    // chr_statistics  = WISECONDORX_PREDICT.out.chr_statistics    // channel: [ val(meta), path(txt) ]
    // chr_plots       = WISECONDORX_PREDICT.out.chr_plots         // channel: [ val(meta), [ path(png), path(png), ... ] ]
    // genome_plot     = WISECONDORX_PREDICT.out.genome_plot       // channel: [ val(meta), path(png) ]

    // versions        = ch_versions                               // channel: path(versions.yml)
}


// Select dorado model
// Dorado models can be selected by auto selection (e.g. hac selects the latest compatible hac model) or by direct model selection.
// In the case of direct selection, we need to add the container path to the model.
dorado_model = params.dorado_auto_model
if(params.dorado_model) {
    dorado_model = "/home/" + params.dorado_model
}

// Check bc-kit
def dorado_bc_kit = null
if(params.dorado_bc_kit && !(params.dorado_bc_kit in params.bc_kits)) {
    exit 1, "Invalid barcode kit specified: ${params.dorado_bc_kit}"
}
// Extract barcode kit value from the summary file
if (!params.dorado_bc_kit) {
    // Find the summary file
    def summaryFileDir = new File(params.run_dir)
    def summaryFileName = summaryFileDir.list().find { it.contains('final_summary') && it.endsWith('.txt') }
    // Check if a summary file was found
    if (!summaryFileName) {
        exit 1, "No final summary file found in ${summaryFileDir}."
    }
    // Create a File object from the filename
    def summaryFile = new File(summaryFileDir, summaryFileName)
    // Read the entire content of the summary file as a single string
    def summaryContent = summaryFile.readLines().join('\n')

    // Find the first matching barcode kit in the summary file
    def extrapolatedBcKit = params.bc_kits.find { bc_kit ->
        summaryContent.contains(bc_kit)
    }

    // Set `params.dorado_bc_kit` to the found kit or null if no match
    dorado_bc_kit = extrapolatedBcKit ?: null
}

// Extract run_id
def runid = file(params.run_dir).name
