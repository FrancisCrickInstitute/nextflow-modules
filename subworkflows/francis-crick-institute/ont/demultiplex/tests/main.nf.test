nextflow_process {

    name "Test Subworkflow ONT_DEMULTIPLEX"
    script "../main.nf"
    process "ONT_DEMULTIPLEX"

    tag "subworkflows"
    tag "ont"
    tag "ont_demux"

    test("simplex_default") {
        tag "ont_demux/simplex_default"

        when {
            process {
                """
                input[0] = "hac"
                input[1] = null
                input[2] = false
                input[3] = null
                input[4] = 999999
                input[5] = '$baseDir/' + params.test_data_dir + '/ont/seq_runs/pod5_100'
                input[6] = null
                input[7] = null
                input[8] = false
                input[9] = '$baseDir/' + params.test_data_dir + '/ont/seq_runs/pod5_100/samplesheet.csv'
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out.versions).match()},
                { assert process.out.bam.size() == 1 },
                { assert process.out.demux_fastq.size() == 1 },
                { assert process.out.demux_bam.size() == 0 },
            )
        }
    }

    test("simplex_emitbam") {
        tag "ont_demux/simplex_emitbam"

        when {
            process {
                """
                input[0] = "hac"
                input[1] = null
                input[2] = false
                input[3] = null
                input[4] = 999999
                input[5] = '$baseDir/' + params.test_data_dir + '/ont/seq_runs/pod5_100'
                input[6] = null
                input[7] = null
                input[8] = true
                input[9] = '$baseDir/' + params.test_data_dir + '/ont/seq_runs/pod5_100/samplesheet.csv'
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out.versions).match()},
                { assert process.out.bam.size() == 1 },
                { assert process.out.demux_fastq.size() == 0 },
                { assert process.out.demux_bam.size() == 1 },
            )
        }
    }
}

// take:
// val_model         // Specify a short code like HAC or SUP
// val_bc_kit        // The barcode kit to demultiplex against
// val_check_barcode // Specifies if the run directory should be searched for a valid barcode kit
// val_bc_parse_pos  // The parse position to substring the barcode from in the meta data
// val_batch_num     // Number of files in a dorado batch
// val_run_dir       // The string path of the nanopore run directory
// val_bam           // The string path of a BAM file to use instead of basecalling
// val_resume_bam    // The string path of a BAM file to resume basecalling from
// val_emit_bam      // Defines whether demux outputs a bam or fastq file
// val_samplesheet   // The string path of the samplesheet to parse for metadata if given

// emit:
// pod5        = ch_pod5_files  // channel: [ path(pod5) ]
// bam         = ch_bam         // channel: [ val(meta), path(bam) ]
// demux_fastq = ch_demux_fastq // channel: [ val(meta), path(fastq) ]
// demux_bam   = ch_demux_bam   // channel: [ val(meta), path(bam) ]
// versions    = ch_versions    // channel: path(versions.yml)