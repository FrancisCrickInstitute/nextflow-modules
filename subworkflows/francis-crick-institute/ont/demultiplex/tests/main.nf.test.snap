{
    "simplex_emitbam": {
        "content": [
            [
                "versions.yml:md5,0c0dbb3b9a04ec5b9e65bc9d5d9ee1d9",
                "versions.yml:md5,b3930a6150351e7cbdb27f7a749ee8e6"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-02T15:47:15.201334"
    },
    "simplex_default": {
        "content": [
            [
                "versions.yml:md5,0c0dbb3b9a04ec5b9e65bc9d5d9ee1d9",
                "versions.yml:md5,b3930a6150351e7cbdb27f7a749ee8e6"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-02T15:31:03.995769"
    }
}